File: ZigzagConversion.py
def convert(s: str, numRows: int) -> str:
    if numRows == 1:
        return s
    res = ""
    for i in range(0, len(s), 2 * numRows - 2):
        res += s[i]
    if numRows > 2:
        for r in range(1, numRows-1):
            for i in range (r, len(s), 2 * numRows - 2):
                res += s[i]
                if  i + (numRows - 1 - r) * 2 < len(s):
                    res += s[i + (numRows - 1 - r) * 2]
    
    for i in range(numRows-1, len(s), 2 * numRows - 2):
        res += s[i]
    return res

print(convert("PAYPALISHIRING", 3))

print(convert("PAYPALISHIRING", 4))

print(convert("A", 1))

File: leetcode/ZigzagConversion.cpp
#include <string>
#include <iostream>

using namespace std;

// class Solution {
// public:
    string convert(string s, int numRows) {
        if (numRows == 1) return s;
        char tmp[numRows][s.length()];
        memset(tmp, 0, sizeof tmp);
        int curR = 0, curC = 0, curIndex = 0;
        bool isDown = true;

        while (curIndex < s.length()) {
            tmp[curR][curC] = s[curIndex];
            ++curIndex;
            if (isDown) {
                if (curR < numRows-1) {
                    ++curR;
                } else {
                    ++curC;
                    --curR;
                    isDown = false;
                }
            } else {
                if (curR > 0) {
                    --curR;
                    ++curC;
                } else {
                    ++curR;
                    isDown = true;
                }
            }
        }

        string res = "";
        for (int r = 0; r < numRows; ++r) {
            for (int c = 0; c < s.length(); ++c) {
                if (tmp[r][c] != '\0') {
                    res += tmp[r][c];
                }
            }
        }
        return res;
    }
// };

int main() {
    cout << convert("PAYPALISHIRING", 3) << endl;
    cout << convert("PAYPALISHIRING", 4) << endl;
    cout << convert("A", 1) << endl;
    return 0;
}